docs/apworld dev faq.md

```diff
@@ -0,0 +1,30 @@
+# APWorld Dev FAQ
+
+This document is meant as a reference tool to show the best known solutions to common problems when developing an apworld
+
+---
+
+### My game has a restrictive start that leads to fill errors
+
+Hint to the Generator that an item needs to be in sphere one with local_early_items
+```
+early_item_name = "Sword"
+self.multiworld.local_early_items[self.player][early_item_name] = 1
+```
+
+---
+
+### I have multiple settings that change the item/location pool counts and need to balance them out
+
+Create extra filler based on the difference between your unfilled locations and your itempool by comparing [get_unfilled_locations](/BaseClasses.py) to your list of items to submit
+
+Note: to use self.create_filler(), self.get_filler_item_name() should be defined to only return valid filler item names
+```
+total_locations = len(self.multiworld.get_unfilled_locations(self.player))
```

Comment:
I don't like advertising this solution in a document like this.

I'm not saying that this code is bad or that worlds shouldn't be doing this - just that it's not ideal enough to be advertising it in this document.

This looks like a workaround for having implemented something else in a less-than-ideal way.

In general I don't like the pattern of world X asking the multiworld what world X did (asking "How many locations did I create?")
Ideally, world X should already know what world X did because world X is world X.

This might be the best workaround after having made lots of other choices that led to this situation.
But the problem is that people will read this, seeing it as an ideal solution, before they have made those other choices that led to this situation.

Reply:
Personally, I don't see just asking the multiworld how many locations are unfilled as being an issue.
But, if you do see an issue, please suggest a specific change here.

If the standard for the document is "only the most ideal solution", then we might as well just close this PR now, imo. No one will ever agree on the most ideal solution for most things that would be useful to have in a doc like this.

Reply:
> If the standard for the document is "only the most ideal solution"

This is a misrepresentation of what I'm saying and a strawman.
"it's not ideal enough" does not mean, imply, or suggest "only the most ideal"

Reply:
the issue states "I have multiple settings that change the item/location pool counts and need to balance them out"
so it's already under the assumption that they have made those design decisions.
But we could always add some extra note/nuance to the header along the lines of "while it is recommended that you build out your locations/items so that the world knows how many locations it created, You can create extra filler based on . . ." 

that aside, some rambles / nitpicks with your argument that essentially amount to: we should probably have a better statement of goal in the document,

saying it's not good enough to be in this document is weird when the point of the document is to have referenceable answers to questions that come up (and this question does come up),
I don't see this snippit as prescriptive - i don't expect everyone to make a new create_non_filler_items() method after reading this. And, albiet its a lot more obvious of an example but i also don't expect everyone to submit "Sword" as an early item.
so showing _a_ way to get location count and more importantly showing them how to use it efficiently makes sense to me

also the idea that a world should know how many locations it submitted and not need to ask the multiworld is strange to me because multiworld.get_locations() (and thus get_unfilled_locations()) already use a cache, so while yes, conceptually its weird to forget what you've already done, it also is weird to expect each individual World to make a better caching system (or alternatively, expect this one FAQ answer to define 1/3 of everyone's World code,,)

Reply:
I wonder how many people are going to find this document when they're starting their world development, and how many people will point to this document for someone just starting their world development,
compared to pointing to / finding this document after a specific question comes up.

If it's more of the latter, I think that makes this example not as much of a problem.
But I don't know which it will be.

My concern is that showing this example to someone who hasn't started making decisions yet about how to create their locations and items, could lead them to making worse decisions about how to create their locations and items.

The specific information used in this example (the number of unfilled locations) is not cached by the multiworld, so this has to take extra time checking to see whether the locations are filled.
(I would think that this was a very good example to show new world developers if it were common for other worlds to put items in my locations before I create my items for the item pool. But that isn't a thing.)

Reply:
> > If the standard for the document is "only the most ideal solution"
> 
> This is a misrepresentation of what I'm saying and a strawman. "it's not ideal enough" does not mean, imply, or suggest "only the most ideal solution"

Please don't call my argument a "strawman". That was not the intent. We should assume good faith in each other, or we should just not interact.

But yes, the wording was a little exaggerated, but my point falls in line with qwint's: This is a fine way to get your unfilled locations, and you even stated as such. The amount of time taken to call `get_unfilled_locations` is probably negligible in a multiworld gen (please correct me if I'm wrong here). Adding some nuance to it, like qwint stated, would be a good idea here. If you have a suggestion for that wording, please share that.

Reply:
> Adding some nuance to it, like qwint stated, would be a good idea here. If you have a suggestion for that wording, please share that.

Maybe something like this:
In an ideal situation your system for producing locations and items wouldn't leave any opportunity for them to be unbalanced. But in real, complex situations, that might be unfeasible.